tsinfer/inference.py

```diff
@@ -1912,6 +1908,146 @@ def get_mutations(self, node_id):
+    then splitting up the grand MRCA (if split_mrca is ``True``), and finally
+    reordering the sample nodes such that they are the first nodes listed in the
+    node table,  removing tree nodes and edges that are not on a path between the
+    root and any of the samples (by applying the :meth:`tskit.TreeSequence.simplify`
+    method with ``keep_unary`` set to True but ``filter_sites``, ``filter_populations``
+    and ``filter_individuals`` set to False).
+
+    :param bool split_mrca: If the oldest node is only a parent to a single
+        "grand MRCA" of all the samples, after removing the oldest node we can
+        split the grand MRCA into separate nodes whenever its children change.
+        The rationale is that a grand MRCA consisting of a single ancestral haplotype
+        with all inference sites in the ancestral state is unlikely to represent a
+        single ancestor in the past. If this parameter is ``False``, any such grand
+        MRCA is *not* split (warning: this may cause problems when further dating the
+        inferred tree sequence). Default: ``True``.
+    """
+    if split_mrca is None:
+        split_mrca = True
+
+    tables = ts.dump_tables()
+
+    if has_single_edge_over_grand_root(ts):
+
+        logger.info("Removing the oldest edge to detach the virtual-root-like ancestor")
+        last_edge = ts.edge(-1)  # Edge with oldest parent is last in the edge table
+        tables.edges.truncate(tables.edges.num_rows - 1)
+
+        # move any mutations above the virtual-root-like ancestor to above the grand
+        # MRCA instead (these will be mutations placed by parsimony)
+        mutations_node = tables.mutations.node
+        mutations_node[mutations_node == last_edge.parent] = last_edge.child
+        tables.mutations.node = mutations_node
+
+        if split_mrca:
+            split_grand_mrca(tables)
+
+    logger.info(
+        "Simplifying with filter_sites=False, filter_populations=False, "
+        "filter_individuals=False, and keep_unary=True on "
+        f"{tables.nodes.num_rows} nodes and {ts.num_edges} edges"
+    )
+    # NB: if this is an inferred TS, match_samples is guaranteed to produce samples
+    # in the same order as passed in to sample_indexes, and simplification will
+    # simply stick all those at the start but keep their order the same
+    tables.simplify(
+        filter_sites=False,
+        filter_populations=False,
+        filter_individuals=False,
+        keep_unary=True,
+    )
+    logger.info(
+        "Finished simplify; now have {} nodes and {} edges".format(
+            tables.nodes.num_rows, tables.edges.num_rows
+        )
+    )
+    return tables.tree_sequence()
+
+
+def split_grand_mrca(tables):
+    # Internal function: if a single oldest node is a root across the entire genome,
+    # split it up into a set of contemporaneous nodes whenever the node children change
+
+    ts = tables.tree_sequence()
+    if not has_same_root_everywhere(ts):
```

Comment:
We haven't done that though, have we? I can only find `has_same_root_everywhere` being called once in the entire file. Above we check whether there is a "virtual-root-like" ancestor, but that's not the same. For instance, this would fail if there were isolated nodes in the TS.

Reply:
Right, sorry, I mixed this up with has_single_edge_over_grand_root

Reply:
Needs a test though, currently not covered.

Reply:
I think it's in `TestRootCheckingFunctions::test_has_same_root_everywhere`?